### PR TITLE
[top-level/pwrmgr] Extend chip_sw_pwrmgr_main_power_glitch_reset

### PR DESCRIPTION
**Pull request by abdullahvarici**
Add two groups of assertions to check that:
- the clock valids are deasserted if a power glitch is detected.
- clocks are stopped if their valid is deasserted.

Signed-off-by: Abdullah Varici <abdullah.varici@lowrisc.org>

### CODE REVIEW

**Comment by abdullahvarici** [comment]
This PR is for issue: https://github.com/lowRISC/opentitan/issues/14780. I am not sure entirely if this is a correct solution.

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv [hunk 1]
```diff
@@ -33,22 +33,38 @@ interface pwrmgr_ast_sva_if (
...
+  always_ff @(posedge clk_usb) usb_clk_cycles++;
+
   // Clock enable-valid.
   `ASSERT(CoreClkHandshakeOn_A, pwr_ast_o.core_clk_en |-> `CLK_WAIT_BOUNDS pwr_ast_i.core_clk_val,
           clk_slow_i, reset_or_disable)
   `ASSERT(CoreClkHandshakeOff_A,
           !pwr_ast_o.core_clk_en |-> `CLK_WAIT_BOUNDS !pwr_ast_i.core_clk_val, clk_slow_i,
           reset_or_disable)
+  `ASSERT(MainClkStopped_A, $fell(pwr_ast_i.core_clk_val) |-> ##[1:2] $stable(main_clk_cycles),
+          clk_slow_i, reset_or_disable)
```
Comment (andreaskurth) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:53]: Isn't this assertion too lax? Would a simple
```sv
pwr_ast_i.core_clk_val |-> $stable(main_clk_cycles)
```
as assertion condition not work?

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv [hunk 2]
```diff
@@ -14,7 +14,6 @@ class chip_sw_main_power_glitch_vseq extends chip_sw_base_vseq;
   virtual task pre_start();
     super.pre_start();
     // disable expected assertion error for power glitch test
-    $assertoff(0,"pwrmgr_ast_sva_if");
```
Comment (andreaskurth) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv:17]: I assume you removed this deliberately, right? If so, could you make it a separate commit and briefly explain why?

**Review by andreaskurth — COMMENTED** [review]
Thanks @abdullahvarici.

To my understanding, the added SVAs cover the checks specified in #14780, but it would be good if @matutem or @tjaychen could confirm.

I've left two comments on the implementation of the assertions.

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv (hunk 1 above)
Comment (abdullahvarici) [reply hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv]: It doesn't work Andreas. I guess we need to check only after falling edge of pwr_ast_i.core_clk_val. But the following also works simply and I've changed to it:
`$fell(pwr_ast_i.core_clk_val) |=> $stable(main_clk_cycles)`

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv (hunk 2 above)
Comment (abdullahvarici) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv:17]: This line was disabling the assertions in pwrmgr_ast_sva_if so I needed to enable that to see to see new added assertions are working. It is related with the other changes, do I really need to create a seperate commit for that?

File: hw/ip/pwrmgr/rtl/pwrmgr.sv [hunk 3]
```diff
@@ -287,6 +287,14 @@ module pwrmgr
   assign hw2reg.fault_status.main_pd_glitch.de  = peri_reqs_masked.rstreqs[ResetMainPwrIdx];
   assign hw2reg.fault_status.main_pd_glitch.d   = 1'b1;
 
+  // Check that the clock valids are deasserted if a power glitch is detected.
```
Comment (matutem) [inline hw/ip/pwrmgr/rtl/pwrmgr.sv:290]: I see, and we expect *clk_en to drop at the latest in the very next slow cycle.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv (hunk 2 above)
Comment (matutem) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv:17]: This disables the assertion specifically for this test. I thought we needed to disable them because they would actually fail given the stimulus we are injecting. If they don't cause failures it is possible it was not necessary to disable them after all..

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv [hunk 4]
```diff
@@ -8,6 +8,9 @@
 interface pwrmgr_ast_sva_if (
   input logic clk_slow_i,
   input logic rst_slow_ni,
+  input logic clk_main_i,
+  input logic clk_io_i,
+  input logic clk_usb_i,
```
Comment (matutem) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:13]: These could cause a failure in the pwrmgr unit test environment.

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv [hunk 5]
```diff
@@ -33,22 +36,36 @@ interface pwrmgr_ast_sva_if (
...
+  always_ff @(posedge clk_io_i) io_clk_cycles++;
+  always_ff @(posedge clk_usb_i) usb_clk_cycles++;
+
   // Clock enable-valid.
   `ASSERT(CoreClkHandshakeOn_A, pwr_ast_o.core_clk_en |-> `CLK_WAIT_BOUNDS pwr_ast_i.core_clk_val,
           clk_slow_i, reset_or_disable)
   `ASSERT(CoreClkHandshakeOff_A,
           !pwr_ast_o.core_clk_en |-> `CLK_WAIT_BOUNDS !pwr_ast_i.core_clk_val, clk_slow_i,
           reset_or_disable)
+  `ASSERT(MainClkStopped_A, $fell(pwr_ast_i.core_clk_val) |=> ($stable(main_clk_cycles)
```
Comment (matutem) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:50]: We should also add the counterpart: when *_val raises the clocks should be ticking.

**Review by matutem — CHANGES_REQUESTED** [review]
This probably breaks the pwrmgr unit tests, since we also bind pwrmgr_ast_sva_if in the unit, and main/usb/io clk_*_i are not inputs to pwrmgr. We may need to add these assertions in a separate *sva_if only used at full chip? Or we could parameterize pwrmgr_ast_sva_if to determine if it is bound for the unit or the full chip and add "if" generates.

In addition, we add "pwrmgr_bind" to "sim_tops" in chip_sim_cfg.hjson, so this would cause these assertions to be bound twice?

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv (hunk 2 above)
Comment (abdullahvarici) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv:17]: Now they don't cause a failure.

File: hw/ip/pwrmgr/rtl/pwrmgr.sv (hunk 3 above)
Comment (abdullahvarici) [reply hw/ip/pwrmgr/rtl/pwrmgr.sv:290]: Sorry, I didn't understand what you mean here.

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv (hunk 5 above)
Comment (abdullahvarici) [reply hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv]: This is also added.

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv (hunk 4 above)
Comment (abdullahvarici) [reply hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:15]: They don't cause failure when using pwrmgr smoke test. I have parameterized pwrmgr_ast_sva_if.

**Comment by abdullahvarici** [comment]
Thanks @matutem for your feedback. Now I've made changes you wanted. Can take a look at it again?

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv [hunk 6]
```diff
@@ -5,9 +5,14 @@
 // This has some assertions that check the inputs from ast react according to
 // the pwrmgr outputs. The ast inputs are generated by the base sequences, but
 // these assertions will also be useful at full chip level.
-interface pwrmgr_ast_sva_if (
+interface pwrmgr_ast_sva_if #(
+  parameter CheckClocks = "False"  // can be either "True" or "False"
```
Comment (andreaskurth) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:9]:
```suggestion
  parameter bit CheckClocks = 1'b0
```

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv [hunk 7]
```diff
@@ -33,6 +38,7 @@ interface pwrmgr_ast_sva_if (
   `define CLK_WAIT_BOUNDS ##[MIN_CLK_WAIT_CYCLES:MAX_CLK_WAIT_CYCLES]
   `define PDN_WAIT_BOUNDS ##[MIN_PDN_WAIT_CYCLES:MAX_PDN_WAIT_CYCLES]
 
+
```
Comment (andreaskurth) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:41]:
```suggestion
```

File: hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv [hunk 8]
```diff
@@ -50,6 +56,43 @@ interface pwrmgr_ast_sva_if (
   `ASSERT(UsbClkHandshakeOff_A, !pwr_ast_o.usb_clk_en |-> `CLK_WAIT_BOUNDS !pwr_ast_i.usb_clk_val,
           clk_slow_i, reset_or_disable)
 
+  if (CheckClocks == "True") begin : gen_check_clock
```
Comment (andreaskurth) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_ast_sva_if.sv:59]:
```suggestion
  if (CheckClocks) begin : gen_check_clock
```

File: hw/top_earlgrey/dv/tb/tb.sv [hunk 9]
```diff
@@ -59,6 +59,20 @@ module tb;
 
   bind chip_earlgrey_asic chip_if chip_if();
 
+  bind dut pwrmgr_ast_sva_if #(
+    .CheckClocks("True")
```
Comment (andreaskurth) [inline hw/top_earlgrey/dv/tb/tb.sv:63]:
```suggestion
    .CheckClocks(1'b1)
```

File: hw/top_earlgrey/dv/tb/tb.sv [hunk 10]
```diff
@@ -59,6 +59,20 @@ module tb;
...
+  ) pwrmgr_ast_sva_if (
+    .clk_slow_i(`PWRMGR_HIER.clk_slow_i),
+    .rst_slow_ni(`PWRMGR_HIER.rst_slow_ni),
+    .clk_main_i(top_earlgrey.clk_main_i),
+    .clk_io_i(top_earlgrey.clk_io_i),
+    .clk_usb_i(top_earlgrey.clk_usb_i),
```
Comment (andreaskurth) [inline hw/top_earlgrey/dv/tb/tb.sv:69]:
```suggestion
    .clk_main_i(`TOP_HIER.clk_main_i),
    .clk_io_i(`TOP_HIER.clk_io_i),
    .clk_usb_i(`TOP_HIER.clk_usb_i),
```

File: hw/ip/pwrmgr/dv/sva/pwrmgr_bind.sv [hunk 11]
```diff
@@ -34,9 +34,14 @@ module pwrmgr_bind;
     .usb_clk_en(pwr_ast_o.usb_clk_en)
   );
 
-  bind pwrmgr pwrmgr_ast_sva_if pwrmgr_ast_sva_if (
+  bind pwrmgr pwrmgr_ast_sva_if #(
+    .CheckClocks("False")
```
Comment (andreaskurth) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_bind.sv:38]:
```suggestion
    .CheckClocks(1'b0)
```

File: hw/ip/pwrmgr/rtl/pwrmgr.sv (hunk 3 above)
Comment (andreaskurth) [reply hw/ip/pwrmgr/rtl/pwrmgr.sv:290]: I think @matutem is slightly rephrasing the comment you added to confirm the understanding. Maybe you could factor the comment in like this:
```suggestion
  // Check that the clock enables are deasserted in the next slow clock cycle if a power glitch is detected.
```

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv (hunk 2 above)
Comment (andreaskurth) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv:17]: Then I think this is resolved

**Review by andreaskurth — CHANGES_REQUESTED** [review]

**Comment by andreaskurth** [comment]
The failing CI checks look spurious to me. They should go away on the next run.

**Comment by abdullahvarici** [comment]
Thanks @andreaskurth for your suggestions. I've added them.

File: hw/ip/pwrmgr/dv/sva/pwrmgr_bind.sv [hunk 12]
```diff
@@ -34,9 +34,14 @@ module pwrmgr_bind;
...
 
-  bind pwrmgr pwrmgr_ast_sva_if pwrmgr_ast_sva_if (
+  bind pwrmgr pwrmgr_ast_sva_if #(
+    .CheckClocks(1'b0)
+  ) pwrmgr_ast_sva_if (
     .clk_slow_i,
     .rst_slow_ni,
+    .clk_main_i(1'bx),
+    .clk_io_i(1'bx),
+    .clk_usb_i(1'bx),
```
Comment (msfschaffner) [inline hw/ip/pwrmgr/dv/sva/pwrmgr_bind.sv:44]: Why are these tie to `x` needed? It would be good to add a comment.

File: hw/top_earlgrey/dv/tb/tb.sv [hunk 13]
```diff
@@ -59,6 +59,26 @@ module tb;
 
   bind chip_earlgrey_asic chip_if chip_if();
 
+`ifdef XCELIUM
+  `define TOP_HIER          tb.dut.top_earlgrey
+`else
+  `define TOP_HIER          top_earlgrey
+`endif
+
+  bind dut pwrmgr_ast_sva_if #(
```
Comment (msfschaffner) [inline hw/top_earlgrey/dv/tb/tb.sv:68]: @sriyerg can probably give more guidance on this, but I thought the idea is to make the IP-level bind file (`pwrmgr_bind` in this case) generic enough so that they can be reused at the top-level.

In fact, looking at the top-level simulation config, I think the `pwrmgr_bind` is already hooked into the top-level simulation here:
https://github.com/lowRISC/opentitan/blob/da1c2ed8171934c90d44a3b6f56e2c19b06a268d/hw/top_earlgrey/dv/chip_sim_cfg.hjson#L26-L34

so this instantiation here is likely not needed.

**Review by msfschaffner — COMMENTED** [review]
Thanks @abdullahvarici - I have some comments, but I believe they are along the lines of what @matute already pointed out.

File: hw/top_earlgrey/dv/tb/tb.sv (hunk 13 above)
Comment (abdullahvarici) [reply hw/top_earlgrey/dv/tb/tb.sv:68]: Thanks @msfschaffner for feedback. Now pwrmgr_ast_sva_if interface has some clk_*_i ports to be used in some assertions. But they are used for top level tests but not used for unit tests. To provide that I also added a parameter to the interface. But in order to do that, I instantiated a new bind in tb.sv and removed IP-level bind file usage from chip_sim_cfg.hjson. I am not sure how to use same bind file for unit and top level tests in our scenario. Also should I remove bind to 1'bx in hw/ip/pwrmgr/dv/sva/pwrmgr_bind.sv?

**Comment by abdullahvarici** [comment]
I tried to address your feedback (@matutem and @msfschaffner ) and it would be great you could take a look again. Also, I am happy to implement further feedback.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv (hunk 2 above)
Comment (matutem) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_main_power_glitch_vseq.sv:17]: Cool

File: hw/top_earlgrey/dv/chip_sim_cfg.hjson [hunk 14]
```diff
@@ -25,7 +25,6 @@
 
   // Add additional tops for simulation.
   sim_tops: ["clkmgr_bind",
-             "pwrmgr_bind",
```
Comment (matutem) [inline hw/top_earlgrey/dv/chip_sim_cfg.hjson:28]: Please don't remove pwrmgr_bind: there are more assertions than pwrmgr_ast, so you are removing more checks. Instead you could rename dv/sva/pwrmgr_rstmgr_bind.sv to something like pwrmgr_unit_bind.sv and place the version of pwrmgr_ast_sva_if suitable for unit testing (with no clocks) there. That way we can continue to bind pwrmgr_bind in top-level with the other assertions.

Then I would put the pwrmgr_ast_sva_if suitable for top-level in top_earlgrey_bind. This will make it easier to understand the flow, since it would be more like the way we handle pwrmgr_rstmgr_sva_if which has a similar discrepancy between top and unit level.

File: hw/top_earlgrey/dv/tb/tb.sv (hunk 13 above)
Comment (matutem) [reply hw/top_earlgrey/dv/tb/tb.sv:68]: See my comment on chip_sim_cfg.hjson.

**Review by matutem — CHANGES_REQUESTED** [review]

**Comment by abdullahvarici** [comment]
I made changes according to your feedback. Could review it again? @matutem

**Comment by msfschaffner** [comment]
This may need a rebase onto master to fix the FPGA CI failures.

**Comment by abdullahvarici** [comment]
Thanks @msfschaffner, I've rebased it.

**Review by matutem — APPROVED** [review]
Thanks for your patience :)
LGTM
